File: OpenCV演算法精解：基於Python與C++/code/unzip/OpenCV算法精解-代码/第9章：几何形状的检测和拟合/9.4/convexityDefects.py
# -*- coding: utf-8 -*-
import cv2
import numpy as np
#轮廓
contour = np.array([[20,20],[50,70],[20,120],[120,120],[100,70],[120,20]],np.int32)
#轮廓的凸包
hull = cv2.convexHull(contour,returnPoints=False)
defects = cv2.convexityDefects(contour,hull)
#打印凸包
print hull
#打印凸包的缺陷
print defects

File: OpenCV演算法精解：基於Python與C++/code/unzip/OpenCV算法精解-代码/第9章：几何形状的检测和拟合/9.4/convexityDefects.cpp
#include<opencv2/core.hpp>
#include<opencv2/imgproc.hpp>
#include<opencv2/highgui.hpp>
using namespace cv;
#include<iostream>
using namespace std;
int main(int argc, char*argv[])
{
	//轮廓
	vector<Point> contour;
	contour.push_back(Point(20, 20));
	contour.push_back(Point(50, 70));
	contour.push_back(Point(20, 120));
	contour.push_back(Point(120, 120));
	contour.push_back(Point(100, 70));
	contour.push_back(Point(120, 20));
	//计算轮廓的凸包
	vector<int> hull;
	convexHull(contour, hull, false, false);
	//计算凸包的缺陷
	vector<Vec4i> defects;
	convexityDefects(contour, hull, defects);
	//打印凸包的缺陷
	for (int i = 0; i < defects.size(); i++)
	{
		cout << defects[i] << endl;
	}
	return 0;
}